### PR TITLE
Add SourceBuffer to deal with Unicode encoding

### PR DESCRIPTION
**Pull request by zrlk**
Even if we weren't using a parser that deals in UTF-16
code units, we'd still have to tolerate source maps that
do. (I've experimentally verified that Mozilla's source
map library generates colums in code units, not code points
(or whatever else you can imagine). The source map standard
is mum on this point.) SourceBuffer maps between UTF-8
code unit offsets and UTF-16 code unit offsets (and 0-based
lines and columns, and source map segments).

### CODE REVIEW

File: anodyne/base/source_buffer.cc [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+  }
+  return utf8_offset;
+}
+
+int SourceBuffer::OffsetForUtf16Offset(int offset) const {
+  auto i = std::lower_bound(line_to_utf16_offset_.begin(),
+                            line_to_utf16_offset_.end(), offset);
+  // *i >= offset || no such *i
+  if (i == line_to_utf16_offset_.end()) {
+    if (line_to_utf16_offset_.empty() || line_to_offset_.empty()) return -1;
```
Comment (creachadair) [inline anodyne/base/source_buffer.cc]: While it's perfectly correct here, I suggest doing this check before the lower_bound comparison. It doesn't really related to _i_ at all.

File: anodyne/base/source_buffer_test.cc [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+constexpr char kUtf82B[] =
+    R"({"version":3,"sources":["utf8_2b"],"names":["r","u2"],"mappings":"AAAAA,CAAC,CAACC,EAAE,CAACD,CAAC;AACN","file":"utf8_2b.map"})";
+constexpr char kUtf83B[] =
+    R"({"version":3,"sources":["utf8_3b"],"names":["r","u3"],"mappings":"AAAAA,CAAC,CAACC,EAAE,CAACD,CAAC;AACN","file":"utf8_3b.map"})";
+constexpr char kUtf84B[] =
+    R"({"version":3,"sources":["utf8_4b"],"names":["r","u4"],"mappings":"AAAAA,CAAC,CAACC,GAAE,CAACD,CAAC;AACN","file":"utf8_4b.map"})";
+constexpr char kAsciiOnlyGen[] = R"(S1Big
+text text
+S2Big)";
+constexpr char kUtf162UGen[] = "s x s";
```
Comment (creachadair) [inline anodyne/base/source_buffer_test.cc]: I think it'd be easier to understand the relationship between the generated maps and these source forms if the sources came first, and if the comment mentioned the relationship.

File: anodyne/tools/source_map_test_gen/index.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
+var fs = require("fs")
```
Comment (creachadair) [inline anodyne/tools/source_map_test_gen/index.js]: Please add a header comment saying what this file is doing.

File: anodyne/base/source_buffer.cc [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+// 0x80     |    0x7FF | 110xxxxx | 10xxxxxx | -        | -        |
+// 0x800    |   0xFFFF | 1110xxxx | 10xxxxxx | 10xxxxxx | -        |
+// 0x10000  | 0x10FFFF | 11110xxx | 10xxxxxx | 10xxxxxx | 10xxxxxx |
+
+constexpr int kTwoByteMark = 0xC0;
+constexpr int kThreeByteMark = 0xE0;
+constexpr int kFourByteMark = 0xF0;
+constexpr int kContinuationMark = 0x80;
+
+int ReadUtf8(const unsigned char* buffer, int pos, int bound, int* out_pos) {
```
[the remark this answers is not in the document]
Comment (zrlk) [reply anodyne/base/source_buffer.cc]: done

File: anodyne/base/source_buffer.cc [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+// UTF-16 encoding, via https://en.wikipedia.org/wiki/UTF-16
+// 0x0000-0xD7FF and 0xE000-0xFFFF => pass-through as one code unit
+// 0xD800-0xDFFF => reserved for surrogate encoding
+// 0x10000-0x10FFFF => for code point c:
+//   c -= 0x10000
+//   high surrogate is (c >> 10) + 0xD800
+//   low surrogate is (c & 0x3FF) + 0xDC00
+int Utf16CodeUnitsFor(int code_point) { return code_point >= 0x10000 ? 2 : 1; }
```
[the remark this answers is not in the document]
Comment (zrlk) [reply anodyne/base/source_buffer.cc]: done

File: anodyne/base/source_buffer.h [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  int OffsetForUtf16LineCol(int line, int col) const;
+
+  /// \param offset file offset in UTF-16 units (not code points!)
+  /// \return -1 for out of bounds; otherwise the byte offset for offset in
+  /// this file's source encoding.
+  int OffsetForUtf16Offset(int offset) const;
+
+  /// \param offset file offset in bytes
+  /// \return null for no mapping; otherwise, the `SourceMapSegment` located
+  /// at `offset`.
```
[the remark this answers is not in the document]
Comment (zrlk) [reply anodyne/base/source_buffer.h]: done

File: anodyne/base/source_buffer.h [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+/// indices?).
+///
+/// Line and column numbers are always zero-based.
+class SourceBuffer {
+ public:
+  SourceBuffer(std::string&& content, SourceMap&& source_map);
+  SourceBuffer(absl::string_view content, SourceMap&& source_map)
+      : SourceBuffer(std::string(content), std::move(source_map)) {}
+
+  /// \param line 0-based line number.
```
[the remark this answers is not in the document]
Comment (zrlk) [reply anodyne/base/source_buffer.h]: The source map spec is 1-based; I normalize *everything* to 0-based.

File: anodyne/base/source_buffer.cc (hunk 1 above)
Comment (zrlk) [reply anodyne/base/source_buffer.cc]: What do you mean? Add another set of tests to check if `line_to_utf16_offset_.empty()` or `offset > line_to_utf16_offset_.back()`? 

I've clarified the comment.

File: anodyne/tools/source_map_test_gen/index.js (hunk 3 above)
Comment (zrlk) [reply anodyne/tools/source_map_test_gen/index.js]: done

File: anodyne/base/source_buffer.cc [hunk 8]
```diff
@@ -35,6 +35,13 @@ constexpr int kThreeByteMark = 0xE0;
...
 constexpr int kContinuationMark = 0x80;
 
+/// \brief Reads a single UTF-8 code point.
+/// \param buffer the buffer to read from.
+/// \param pos the position in the buffer.
+/// \param bound the length of the buffer.
+/// \param out_pos set to the position immediately following the code point
+/// at pos.
+/// \return the value of the code point at pos.
 int ReadUtf8(const unsigned char* buffer, int pos, int bound, int* out_pos) {
```
Comment (creachadair) [inline anodyne/base/source_buffer.cc]: Is `int` the right type here? I'd think you'd at least want `int32_t` to be sure you have enough bits?

Comment (zrlk) [reply anodyne/base/source_buffer.cc]: done

**Pull request merged**